sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala

```diff
@@ -307,11 +307,12 @@ private[jdbc] class JDBCRDD(
       ""
     }
 
+    val myTopExpression: String = dialect.getTopExpression(limit) // SQL Server Limit alternative
     val myLimitClause: String = dialect.getLimitClause(limit)
     val myOffsetClause: String = dialect.getOffsetClause(offset)
 
     val sqlText = options.prepareQuery +
-      s"SELECT $columnList FROM ${options.tableOrQuery} $myTableSampleClause" +
+      s"SELECT $myTopExpression $columnList FROM ${options.tableOrQuery} $myTableSampleClause" +
```

Comment:
It's hard to define a SQL template that works for all dialects. The approach we take is to try to make it a superset of all dialects.

A better design is to let the dialect build the SQL. For example, we can add a new API in `JdbcDialect`
```
def getSQLBuilder: SQLQueryBuilder = ... the default implementation builds the SQL in a standard way.
```
Then Spark gets the builder and do
```
val builder = ...
builder.withColumns(...).withFromRelation(...).withLimit(...)...build()
```

SQLServerDialect can put the limit before the column list.

Reply:
I will do the work.